### PR TITLE
handle type compatibilty for 4.x baseline

### PR DESCRIPTION
**Pull request by psycofdj**
This PR provides support for older version of the 4.x API.
Some fields like `managementserverid` were previously serialized as `long` and are now serialized as `string`. 

This type difference makes the client fail on response parsing with errors like `json: cannot unmarshal number into Go struct field Host.host.managementserverid of type string"`.

Since these kind of unexpected type changes are frequent in cloudstack API (within a single major version), the PR introduces a static list of fields that should be parsed as either `long` or `string`

### CODE REVIEW

**Comment by rhtyd** [comment]
@psycofdj Why not just upgrade your library/fix clients? String is more easily serializable than long wrt compatibilities?

**Comment by psycofdj** [comment]
You mean upgrading cloudstack manager in my production environment ? If so, I wish I could. Unfortunately, this is not in my hands and, from what I heard, has a strong dependency on underlying operating system. 

I totally agree with you that string are much more easy to use than long. However my current cloudstack 4.11.3 is giving my longs for this field and there is nothing I can do about it. Because newer versions of cloudstack (ie: 4.15) are using strings, my PR implements a flexible parsing that handle both types.

The root cause of this problem is the cloudstack API's stability. In my opinion, these kind of type changes should have not occurred within a given major version of the API.

**Comment by rhtyd** [comment]
Nope, that's not what I meant @psycofdj you only need to fix your apps that are using go-sdk. The move to the field from long to string would make it compatible (for older ACS versions the mgmt server ID field will be returned as string than long).

I think all you need to do is to try/test rebuild your go-based app/code which uses the CloudStack go-sdk.

**Comment by psycofdj** [comment]
I still don't totally understand what your suggestion is (please pardon me for this ^^). Allow me to try a different interpretation and tell me if I'm getting closer:

- I need to modify my golang application that are using `cloudstack-go's` bindings to support `strings` instead of `long` in my code for the fields for which I have problems. 
  - the problem is not the type representation in my applications. The problem is the json parsing of cloudstack's responses coded in `cloudstack-go`. 
  - In golang, the json parser requires a perfect match between types given in the json object and the types described in the receiving json structure. 
  - In my case, cloudstack 4.11.3 if giving me `managementserverid` field as a json string `{"managementserverid":  123456789}`, and the `cloudstack-go` sdk [wants it as a string](https://github.com/apache/cloudstack-go/blob/main/cloudstack/HostService.go#L360), which lead to a parsing error that cannot be recovered by my application using the sdk.

I'm using this sdk in various golang applications related to deploying running Cloudfoundry on cloudstack:
- [orange-cloudfoundry/bosh-cpi-cloudstack](https://github.com/orange-cloudfoundry/bosh-cpi-cloudstack) : cloudstack driver for [BOSH](http://bosh.io/)
- [orange-cloudfoundry/terraform-provider-cloudstack](https://github.com/orange-cloudfoundry/terraform-provider-cloudstack): cloudstack driver for [terraform](https://www.terraform.io/)
- other small close-source projects

**Comment by psycofdj** [comment]
To put it in another way: to be able to read my cloudstack's response, I **need** to modify somehow the code of the cloudstack-go-sdk I'm using.
The funny part is that I don't even use this particular field, I need the response to be parsed.

**Comment by rhtyd** [comment]
Thanks for sharing @psycofdj I understand your problem, you can do two things:
- Pin your projects to use same go-sdk tag which works with your cloudstack version (in go.mod etc) https://github.com/apache/cloudstack-go/tags
- Handle the case of mgmt server ID in main branch so it would serialise both integer and string IDs as string in the struct (so json->struct won't break)

**Comment by psycofdj** [comment]
Unfortunately, there is no available tags in [https://github.com/apache/cloudstack-go/tags](github.com/apache/cloudstack-go) that matches this cirteria. We already had to fork the project when this repo was in xanzy's namespace.

This leave the solution 2. which is exactly what this PR is implementing.

File: cloudstack/HostService.go [hunk 1]
```diff
@@ -357,7 +357,7 @@ type AddBaremetalHostResponse struct {
 	Jobstatus                        int                                `json:"jobstatus"`
 	Lastannotated                    string                             `json:"lastannotated"`
 	Lastpinged                       string                             `json:"lastpinged"`
-	Managementserverid               string                             `json:"managementserverid"`
+	Managementserverid               CSLong                             `json:"managementserverid"`
```
Comment (rhtyd) [inline cloudstack/HostService.go:360]: Could something like this work?
```suggestion
	Managementserverid               string                             `json:"managementserverid,string"`
```

Comment (psycofdj) [reply cloudstack/HostService.go:360]: I'm not familiar with this construction but from my understanding it won't work.
- first: [encoding/json documentation](https://pkg.go.dev/encoding/json#Marshal) states that this tags use use to tell the library to store the value as a json-encoded string. It cannot be use to read an json-long as a string
- second: I created a test program on golang's playgournd that shows that, indeed, this construction wont work. https://play.golang.org/p/iS3FVwp7mik

Comment (psycofdj) [reply cloudstack/HostService.go:360]:
```go
package main

import (
	"fmt"
	"encoding/json"
)

type test struct {
  Test string `json:"test,string"`
}

func main() {
	val := `{"test": 12564564}`
	t := test{}
	err := json.Unmarshal([]byte(val), &t)
	if err != nil {
		fmt.Printf("error: %s\n", err)
	} else {
		fmt.Printf("%#v\n", t)
	}
}
```
`error: json: invalid use of ,string struct tag, trying to unmarshal unquoted value into string`

**Comment by davidjumani** [comment]
@psycofdj The v2.10.0 and lower tag have int64 for the Managementserverid. I'd prefer having tags for each version rather than partial backward compatibility

**Comment by psycofdj** [comment]
@davidjumani are you suggesting to maintain multiple `v2.x.y` branches and version where each `x.y` will follow one specific version of cloudstack's api ?

**Comment by psycofdj** [comment]
In the current state I can't use tags lower than v2.10.0 because of other problems, typically fixed by #7 and #19

**Comment by psycofdj** [comment]
@davidjumani any idea on what should I do with this PR ?

**Comment by psycofdj** [comment]
rebased for conflicts

**Comment by davidjumani** [comment]
I'm not for this since it introduces a new type it can break existing integrations
`cannot use resp.HostsMetrics[1].Managementserverid (type cloudstack.CSLong) as type string in assignment`

**Comment by rhtyd** [comment]
Can we do something like PL has done recently
https://github.com/apache/cloudstack-go/commit/d0a8e7d2a3623d00911b71b5e9ce514487658d6e @davidjumani review and propose a PR to fix the issue and maintain compatibility?

**Comment by psycofdj** [comment]
> cannot use resp.HostsMetrics[1].Managementserverid (type cloudstack.CSLong) as type string in assignment
I'm surprised that this error occurs. Can you provide a sample of the input json that creates this error ?

**Comment by davidjumani** [comment]
The input json isn't the error, it's the new type. The Managementserverid used to be a string and with this patch is now a CSLong. So trying to use the new type where a string used to be causes the error 

```
var msID string
...............
resp, err := cs.Host.ListHostsMetrics()
msID = resp.HostsMetrics[1].Managementserverid     <- resp.HostsMetrics[1].Managementserverid (type cloudstack.CSLong) as type string in assignment
```

**Comment by psycofdj** [comment]
That is why I'm surprised. The goal of this new type is to handle **both** string (latest versions of CS api) and **long** (previous versions of CS api). With this type, it should have parsed correctly your json input.

I'm aware that the json is not the probem, I was wondering if you could show me part of it so that I can test and understand why the `CSLong` type did not parse it correctly.

**Comment by psycofdj** [comment]
@davidjumani I've spotted a problem with the parsing, can you retest with the latest commit ?

**Comment by davidjumani** [comment]
This does handle and parse both int and string, but since the type of the Managementserverid has changed from string to CSLong, it can no longer be used as it was before, but rather needs to be explicitly converted to string which can break existing applications

In the following example, the `msID` field is a string that is then used to store the `Managementserverid` and used later on in the program. Due to the change in the type of Managementserverid, it now throws the error or needs explicit conversion before it can be used to store the value of Managementserverid

```
var msID string
...............
resp, err := cs.Host.ListHostsMetrics()
msID = resp.HostsMetrics[1].Managementserverid     <- resp.HostsMetrics[1].Managementserverid (type cloudstack.CSLong) as type string in assignment

msID = string(resp.HostsMetrics[1].Managementserverid)     <- works
```

**Comment by davidjumani** [comment]
That being said, I'm a +0 on this although I'd prefer if the name of the new data type is more intuitive such as UUID, LongID, etc

**Comment by psycofdj** [comment]
I now understand your problem and, unfortunately,  I have no solution to this.

So, want should we do now ? What are the other solutions to handle past and future type breaking changes in the cloudstack API ?

**Comment by rhtyd** [comment]
Considering all the discussions we've had, I think it's safe to say that moving forward the specific datatype of the mgmt server ID is string. The only cons are that whoever may be using this response attribute will have to handle the typecast of how it is consumed by their code, however, this is not something most people would use out of an API request. I think it makes sense to rename the struct to something like ManagementServerID or UUID and have the changes such that at least it can handle all past/new ACS releases wherein the response object was integer and string.

**Comment by davidjumani** [comment]
@psycofdj I'm unable to edit this PR, so created another one with these commits along with the type name change from CSLong to ManagementServerId in #23

**Comment by psycofdj** [comment]
Alright I shall close this PR then. Thanks @davidjumani

**Pull request closed without merging**